### PR TITLE
[flang][OpenMP] Implement copyin for pointers and allocatables.

### PR DESCRIPTION
**Pull request by DavidTruby**
The copyin clause currently forbids pointer and allocatable variables,
which are allowed by the OpenMP 1.1 and 3.0 specifications respectively.

For pointer variables it is sufficient to remove the check to get
correct behaviour.

For allocatables we need to ensure that the master thread allocatable is
assigned to all of the threads, so we add an hlfir.assign.

### CODE REVIEW

File: flang/lib/Lower/OpenMP/ClauseProcessor.cpp [hunk 1]
```diff
@@ -619,9 +620,28 @@ bool ClauseProcessor::processCopyin() const {
...
+            assert(sb && "Host-associated symbol box not found");
+            assert(hsb.getAddr() != sb.getAddr() &&
+                   "Host and associated symbol boxes are the same");
+
+            mlir::Location loc = converter.genLocation(sym->name());
+            mlir::OpBuilder::InsertionGuard ipGuard{builder};
+            builder.setInsertionPointAfter(sb.getAddr().getDefiningOp());
+            builder.create<hlfir::AssignOp>(loc, hsb.getAddr(), sb.getAddr(),
+                                            true, false, false);
+          }
```
Comment (luporl) [inline flang/lib/Lower/OpenMP/ClauseProcessor.cpp:643]: Can't `copyHostAssociateVar` be used instead?

After https://github.com/llvm/llvm-project/pull/106559, it should be easier to use it to this end and avoid code duplication.

Comment (DavidTruby) [reply flang/lib/Lower/OpenMP/ClauseProcessor.cpp:643]: `copyHostAssociateVar` is already being called at line 647 but that wasn't sufficient in my testing to mean that the allocatable got copied in. I needed to add the `hlfir.assign` to get the correct behaviour. I was testing with the following program:

```fortran
program main
  integer, allocatable, save :: a
  !$omp threadprivate(a)
  
  !$omp master
  allocate(a)
  a = 1
  !$omp end master
  
  !$omp parallel copyin(a)
  if (allocated(a)) then
    print *, a
  endif
  !$omp end parallel
end program
```

Without the copyin(a) this should only print once, whereas with the copyin(a) it should print OMP_NUM_THREADS times. Without the code block highlighted here, it only prints once.

Comment (luporl) [reply flang/lib/Lower/OpenMP/ClauseProcessor.cpp:643]: The test program works correctly with https://github.com/llvm/llvm-project/pull/106559, if I just remove the TODO:
```
--- a/flang/lib/Lower/OpenMP/ClauseProcessor.cpp
+++ b/flang/lib/Lower/OpenMP/ClauseProcessor.cpp
@@ -619,9 +619,6 @@ bool ClauseProcessor::processCopyin() const {
               checkAndCopyHostAssociateVar(&*mem, &insPt);
             break;
           }
-          if (semantics::IsAllocatableOrObjectPointer(&sym->GetUltimate()))
-            TODO(converter.getCurrentLocation(),
-                 "pointer or allocatable variables in Copyin clause");
           assert(sym->has<semantics::HostAssocDetails>() &&
                  "No host-association found");
           checkAndCopyHostAssociateVar(sym);
```

But line https://github.com/llvm/llvm-project/pull/106559/files#diff-0f70165af00b05acb5abeb11198079b793a20acf8f069edd2dbe7b34ac67b328R1262 should be chaged too, to allow `copyin` with unallocated variables.

Comment (DavidTruby) [reply flang/lib/Lower/OpenMP/ClauseProcessor.cpp:643]: Are you saying, if we remove the todo and also add `|| flags.test(...::CopyIn` in the above place things should work?
When I _just_ removed the TODO it didn't work for me.

Comment (DavidTruby) [reply flang/lib/Lower/OpenMP/ClauseProcessor.cpp:643]: Ah, I didn't have the above patch though. I'll have a look at this tomorrow (it's getting late here)

Comment (luporl) [reply flang/lib/Lower/OpenMP/ClauseProcessor.cpp:643]: > Are you saying, if we remove the todo and also add `|| flags.test(...::CopyIn` in the above place things should work? When I _just_ removed the TODO it didn't work for me.

Yes. I have just merged the patch to main.

Comment (DavidTruby) [reply flang/lib/Lower/OpenMP/ClauseProcessor.cpp:643]: You're right, it's sufficient to just remove the TODO with your patch. Thanks!

**Comment by luporl** [comment]
Looks good, but you also need to add that `|| flags.test(Fortran::semantics::Symbol::Flag::OmpCopyIn)` in `Bridge.cpp`, to handle the copy of unallocated variables. WIthout it the test program below crashes.

```f90
program main
  integer, allocatable, save :: a
  !$omp threadprivate(a)

  !$omp parallel copyin(a)
    a = 1
  !$omp end parallel
end program
```

**Review by luporl — APPROVED** [review]
LGTM, thanks.

**Review by tblah — APPROVED** [review]

**Pull request merged**